kernel/hConstr.ml

```diff
@@ -0,0 +1,370 @@
+    let self = kind_to_constr kind in
+    let self = if hasheq_kind (Constr.kind self) (Constr.kind c) then c else self in
+    let hash = hash_kind kind in
+    let isRel, hash = match kind with
+      | Rel n ->
+        let uid = Range.get local_env.rels (n-1) in
+        assert (uid <> 0);
+        uid, Hashset.Combine.combine uid hash
+      | _ -> 0, hash
+    in
+    { self; kind; hash; isRel; refcount = 1 }
+  in
+  match Tbl.find_opt tbl c with
+  | Some c' -> c'.refcount <- c'.refcount + 1; c'
+  | None -> Tbl.add tbl c c; c
+
+and of_constr_aux tbl local_env c =
+  match kind c with
+  | Var i ->
+    let i = Id.hcons i in
+    Var i
+  | Rel _ as t -> t
+  | Sort s ->
+    let s = Sorts.hcons s in
+    Sort s
+  | Cast (c,k,t) ->
+    let c = of_constr tbl local_env c in
+    let t = of_constr tbl local_env t in
+    Cast (c, k, t)
+  | Prod (na,t,c) ->
+    let na = hcons_annot na in
+    let t = of_constr tbl local_env t in
+    let c = of_constr tbl (push_assum t local_env) c in
+    Prod (na, t, c)
+  | Lambda (na, t, c) ->
+    let na = hcons_annot na in
+    let t = of_constr tbl local_env t in
+    let c = of_constr tbl (push_assum t local_env) c in
+    Lambda (na,t,c)
+  | LetIn (na,b,t,c) ->
+    let na = hcons_annot na in
+    let b = of_constr tbl local_env b in
+    let t = of_constr tbl local_env t in
+    let c = of_constr tbl (push_letin ~body:b ~typ:t local_env) c in
+    LetIn (na,b,t,c)
+  | App (h,args) ->
+    let h = of_constr tbl local_env h in
+    let args = Array.map (of_constr tbl local_env) args in
+    App (h, args)
+  | Evar _ -> CErrors.anomaly Pp.(str "evar in typeops")
+  | Meta _ -> CErrors.anomaly Pp.(str "meta in typeops")
+  | Const (c,u) ->
+    let c = hcons_con c in
+    let u = UVars.Instance.hcons u in
+    Const (c,u)
+  | Ind (ind,u) ->
+    let ind = hcons_ind ind in
+    let u = UVars.Instance.hcons u in
+    Ind (ind,u)
+  | Construct (c,u) ->
+    let c = hcons_construct c in
+    let u = UVars.Instance.hcons u in
+    Construct (c,u)
+  | Case (ci,u,pms,(p,r),iv,c,bl) ->
+    let of_ctx (bnd, c) =
+      let () = hcons_inplace hcons_annot bnd in
+      let local_env = Array.fold_left (fun local_env _ -> push_unknown_rel local_env) local_env bnd in
+      let c = of_constr tbl local_env c in
+      bnd, c
+    in
+    let ci = hcons_caseinfo ci in
+    let u = UVars.Instance.hcons u in
+    let pms = Array.map (of_constr tbl local_env) pms in
+    let p = of_ctx p in
+    let iv = match iv with
+      | NoInvert -> NoInvert
+      | CaseInvert {indices} -> CaseInvert {indices=Array.map (of_constr tbl local_env) indices}
+    in
+    let c = of_constr tbl local_env c in
+    let bl = Array.map of_ctx bl in
+    Case (ci,u,pms,(p,r),iv,c,bl)
+  | Fix (ln,(lna,tl,bl)) ->
+    let () = hcons_inplace hcons_annot lna in
+    let tl = Array.map (of_constr tbl local_env) tl in
+    let body_env = Array.fold_left (fun env t -> push_assum t env) local_env tl in
```

Comment:
This is incorrect with mutual fixpoints and causes the relation_algebra error (the type needs to be lifted)

Reduced bug:
~~~coq
Inductive trip := Z | L (_:trip) (_:trip) | R (_:trip).

Type fun x:bool =>
  fix foo (n:trip) : x = x :=
  match n with
  | Z => eq_refl
  | L k _ => eq_refl
  | R k => bar k
  end
with bar (n:trip) : x = x :=
  match n with
  | Z => eq_refl
  | L k _ => bar k
  | R k => foo k
  end
for foo.
(* Error: In pattern-matching on term "n" the branch for constructor "R" has type
 "fun _ : trip => _UNBOUND_REL_6 = _UNBOUND_REL_6" which should be "fun _ : trip => x = x".
*)
~~~
(the relation_algebra example has a raw fix appear in a subterm through unification)